packages/ninja/src-spawn.cc.patch

```diff
@@ -0,0 +1,387 @@
++union SigSetConverter {
++  int bsd;
++  sigset_t sigset;
++  sigset64_t sigset64;
++};
++
++enum Action {
++  kOpen,
++  kClose,
++  kDup2
++};
++
++struct __posix_spawn_file_action {
++  __posix_spawn_file_action* next;
++
++  Action what;
++  int fd;
++  int new_fd;
++  char* path;
++  int flags;
++  mode_t mode;
++
++  void Do() {
++    if (what == kOpen) {
++      fd = open(path, flags, mode);
++      if (fd == -1) _exit(127);
++      // If it didn't land where we wanted it, move it.
++      if (fd != new_fd) {
++        if (dup2(fd, new_fd) == -1) _exit(127);
++        close(fd);
++      }
++    } else if (what == kClose) {
++      // Failure to close is ignored.
++      close(fd);
++    } else {
++      if (dup2(fd, new_fd) == -1) _exit(127);
++    }
++  }
++};
++
++struct __posix_spawn_file_actions {
++  __posix_spawn_file_action* head;
++  __posix_spawn_file_action* last;
++
++  void Do() {
++    for (__posix_spawn_file_action* action = head; action != nullptr; action = action->next) {
++      action->Do();
++    }
++  }
++};
++
++struct __posix_spawnattr {
++  short flags;
++  pid_t pgroup;
++  sched_param schedparam;
++  int schedpolicy;
++  SigSetConverter sigmask;
++  SigSetConverter sigdefault;
++};
++
++static void ApplyAttrs(short flags, const posix_spawnattr_t* attr) {
++  // POSIX: "If POSIX_SPAWN_SETSIGDEF is set ... signals in sigdefault ...
++  // shall be set to their default actions in the child process."
++  // POSIX: "Signals set to be caught by the calling process shall be
++  // set to the default action in the child process."
++  bool use_sigdefault = ((flags & POSIX_SPAWN_SETSIGDEF) != 0);
++  const struct sigaction default_sa = { .sa_handler = SIG_DFL };
++  for (int s = 1; s < _NSIG; ++s) {
++    bool reset = false;
++    if (use_sigdefault && sigismember(&(*attr)->sigdefault.sigset, s)) {
++      reset = true;
++    } else {
++      struct sigaction current;
++      if (sigaction(s, nullptr, &current) == -1) _exit(127);
++      reset = (current.sa_handler != SIG_IGN && current.sa_handler != SIG_DFL);
++    }
++    if (reset && sigaction(s, &default_sa, nullptr) == -1) _exit(127);
++  }
++
++  if ((flags & POSIX_SPAWN_SETPGROUP) != 0 && setpgid(0, (*attr)->pgroup) == -1) _exit(127);
++  if ((flags & POSIX_SPAWN_SETSID) != 0 && setsid() == -1) _exit(127);
++
++  // POSIX_SPAWN_SETSCHEDULER overrides POSIX_SPAWN_SETSCHEDPARAM, but it is not an error
++  // to set both.
++  if ((flags & POSIX_SPAWN_SETSCHEDULER) != 0) {
++    if (sched_setscheduler(0, (*attr)->schedpolicy, &(*attr)->schedparam) == -1) _exit(127);
++  } else if ((flags & POSIX_SPAWN_SETSCHEDPARAM) != 0) {
++    if (sched_setparam(0, &(*attr)->schedparam) == -1) _exit(127);
++  }
++
++  if ((flags & POSIX_SPAWN_RESETIDS) != 0) {
++    if (seteuid(getuid()) == -1 || setegid(getgid()) == -1) _exit(127);
++  }
++
++  if ((flags & POSIX_SPAWN_SETSIGMASK) != 0) {
++    if (sigprocmask(SIG_SETMASK, &(*attr)->sigmask.sigset, nullptr)) _exit(127);
```

Comment:
Other than inserting two private Bionic headers and a macro above, the only modifications to this upstream implementation of `posix_spawn` were to switch from the 64-bit APIs `sigprocmask64`/`sigset64` to the older Bionic APIs here and a handful of other places, since they aren't available before Pie.